core/routing_dtlsr.go

```diff
@@ -0,0 +1,750 @@
+		graph.AddVertex(i)
+		// log node-index mapping for debug purposes
+		log.WithFields(log.Fields{
+			"index": i,
+			"node":  dtlsr.indexNode[i],
+		}).Debug("Node-index-mapping")
+	}
+
+	// add edges originating from this node
+	for peer, timestamp := range dtlsr.peers.peers {
+		var edgeCost int64
+		if timestamp == 0 {
+			edgeCost = 0
+		} else {
+			edgeCost = int64(currentTime - timestamp)
+		}
+
+		if err := graph.AddArc(0, dtlsr.nodeIndex[peer], edgeCost); err != nil {
+			log.WithFields(log.Fields{
+				"reason": err.Error(),
+			}).Warn("Error computing routing table")
+			return
+		}
+
+		log.WithFields(log.Fields{
+			"peerA": dtlsr.c.NodeId,
+			"peerB": peer,
+			"cost":  edgeCost,
+		}).Debug("Added vertex")
+	}
+
+	// add edges originating from other nodes
+	for _, data := range dtlsr.receivedData {
+		for peer, timestamp := range data.peers {
+			var edgeCost int64
+			if timestamp == 0 {
+				edgeCost = 0
+			} else {
+				edgeCost = int64(currentTime - timestamp)
+			}
+
+			if err := graph.AddArc(dtlsr.nodeIndex[data.id], dtlsr.nodeIndex[peer], edgeCost); err != nil {
+				log.WithFields(log.Fields{
+					"reason": err.Error(),
+				}).Warn("Error computing routing table")
+				return
+			}
+
+			log.WithFields(log.Fields{
+				"peerA": data.id,
+				"peerB": peer,
+				"cost":  edgeCost,
+			}).Debug("Added vertex")
+		}
+	}
+
+	routingTable := make(map[bundle.EndpointID]bundle.EndpointID)
+	for i := 1; i < dtlsr.length; i++ {
+		shortest, err := graph.Shortest(0, i)
+		if err == nil {
+			if len(shortest.Path) <= 1 {
+				log.WithFields(log.Fields{
+					"node_index": i,
+					"node":       dtlsr.indexNode[i],
+					"path":       shortest.Path,
+				}).Warn("Single step path found - this should not happen")
+				continue
+			}
+
+			routingTable[dtlsr.indexNode[i]] = dtlsr.indexNode[shortest.Path[1]]
+			log.WithFields(log.Fields{
+				"node_index": i,
+				"node":       dtlsr.indexNode[i],
+				"path":       shortest.Path,
+				"next_hop":   routingTable[dtlsr.indexNode[i]],
+			}).Debug("Found path to node")
+		} else {
+			log.WithFields(log.Fields{
+				"node_index": i,
+				"error":      err.Error(),
+			}).Debug("Did not find path to node")
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"routingTable": routingTable,
+	}).Debug("Finished routing table computation")
+
+	dtlsr.routingTable = routingTable
+}
+
+// recomputeCron gets called periodically by the core's cron module.
+// Only actually triggers a recompute if the underlying data has changed.
+func (dtlsr *DTLSR) recomputeCron() {
+	log.WithFields(log.Fields{
+		"peerChange":     dtlsr.peerChange,
+		"receivedChange": dtlsr.receivedChange,
+	}).Debug("Executing recomputeCron")
+
+	if dtlsr.peerChange || dtlsr.receivedChange {
+		dtlsr.dataMutex.Lock()
+		dtlsr.computeRoutingTable()
+		dtlsr.receivedChange = false
+		dtlsr.dataMutex.Unlock()
+	}
+}
+
+// broadcast broadcasts this node's peer data to the network
+func (dtlsr *DTLSR) broadcast() {
```

Comment:
Does a broadcast bundle have to be sent epidemically to all network peers? Don't the `dtlsr.peers` also contain transitive neighbors? If so, a hop count block should be added here.

Reply:
In link-state-routing you don't advertise routes (as in distance-vector-routing) but you direct peers. So everyone needs to get this data.